test/test_nn.py

```diff
@@ -4899,26 +4899,45 @@ def helper(self, mod, size, dtype, mixed_dtype=False, format=torch.channels_last
                     helper(self, nn.BatchNorm3d, shape, dtype, mixed_dtype, torch.channels_last_3d, precisons[dtype])
 
     def test_batchnorm_half_overflow(self):
-        def helper(self, mod, size, format):
+        def helper(self, mod, size, param_dtype, fwd_format, bwd_format):
             channels = size[1]
-            input = torch.randn(size, dtype=torch.half, device='cpu', requires_grad=True)
-            input = input.contiguous(memory_format=format)
-            bn = mod(channels).cpu().to(torch.half)
+            input = torch.randn(size, dtype=torch.half, device='cpu')
+            input = input.contiguous(memory_format=fwd_format).requires_grad_(True)
+            bn = mod(channels).cpu().to(param_dtype)
             out = bn(input)
 
+            ref_input = input.detach().clone().requires_grad_(True)
             ref_bn = mod(channels).cpu().to(torch.float)
             ref_bn.load_state_dict(bn.to(torch.float).state_dict())
-            ref_out = ref_bn(input)
+            ref_out = ref_bn(ref_input)
 
             self.assertFalse(out.isinf().any())
             self.assertFalse(out.isnan().any())
             self.assertEqual(out, ref_out)
 
+            if param_dtype != torch.half:
+                grad_input = torch.empty(size=ref_out.shape).uniform_(0, 1).to(dtype=torch.half)
+                grad_input = grad_input.contiguous(memory_format=bwd_format)
+                ref_grad_input = grad_input.clone()
+                out.backward(grad_input)
+                ref_out.backward(ref_grad_input)
+                self.assertFalse(input.grad.isinf().any())
+                self.assertFalse(input.grad.isnan().any())
+                self.assertEqual(input.grad, ref_input.grad)
+
         for format in [torch.contiguous_format, torch.channels_last]:
-            helper(self, nn.BatchNorm2d, (4, 80, 500, 500), format)
+            helper(self, nn.BatchNorm2d, (4, 80, 500, 500), torch.half, format, format)
 
         for format in [torch.contiguous_format, torch.channels_last_3d]:
-            helper(self, nn.BatchNorm3d, (4, 80, 20, 100, 100), format)
+            helper(self, nn.BatchNorm3d, (4, 80, 20, 100, 100), torch.half, format, format)
+
+        for fwd_format in [torch.contiguous_format, torch.channels_last]:
+            for bwd_format in [torch.contiguous_format, torch.channels_last]:
+                helper(self, nn.BatchNorm2d, (16, 3, 224, 224), torch.float, fwd_format, bwd_format)
+
+        for fwd_format in [torch.contiguous_format, torch.channels_last_3d]:
```

Comment:
nit: you could also write this as

```
formats = [torch.contiguous_format, torch.channels_last_3d]
for (fwd_format, bwd_format) in itertools.product(formats, formats):
    helper(...)
```

which should be easier to read.

See the following to illustrate

```
>>> choices = [0, 1]
>>> list(itertools.product(choices, choices))
[(0, 0), (0, 1), (1, 0), (1, 1)]
```

or use `parametrize` as in the other tests to be able to get one test for each combo.

This should make it easier to catch any future failures.

Reply:
Thanks for your comments. Revised.